File: src/daemon/migrator/rabbitmq/rabbitmq.go
package rabbitmq

import (
	"fmt"
	"log"
	"time"

	"github.com/streadway/amqp"
)

var connection *amqp.Connection

func CreateConnection() (*amqp.Connection, error) {
	var (
		user = "is"
		pass = "is"
		vshost = "is"
		connStr = fmt.Sprintf("amqp://%s:%s@broker:5672/%s", user, pass, vshost)
	)
	conn, err := amqp.Dial(connStr)

	if err != nil {
		return nil, err
	}

	return conn, nil
}

func Init() error {
	conn, err := CreateConnection()
	if err != nil {
		return err
	}

	connection = conn

	return nil
}


func ConsumeQueue(queueName string) ([]string, error) {
	timeDuration := 10 * time.Second

    channel, err := connection.Channel()
    if err != nil {
        return nil, err
    }
    defer channel.Close()

    msgs, err := channel.Consume(
        queueName,
        "",
        false,
        false,
        false,
        false,
        nil,
    )
    if err != nil {
		if amqpErr, ok := err.(*amqp.Error); ok && amqpErr.Code == 404 {
			log.Printf("Waiting for %s queue\n", queueName)
			return nil, nil
		}

        return nil, err
    }

	log.Println("Waiting for messages...")
    data := make([]string, 0)

    go func() {
        for msg := range msgs {
			messageBody := string(msg.Body)
            data = append(data, messageBody)
			msg.Ack(false)
        }
    }()

	time.Sleep(timeDuration)
	if len(data) != 0 {
		log.Println("Done receiving messages.")
	}

    return data, nil
}

func AddToQueue(queueName string, messageBody string) error {
    channel, err := connection.Channel()
    if err != nil {
        return err
    }
    defer channel.Close()

    queue, err := channel.QueueDeclare(
        queueName,
        false,
        false,
        false,
        false,
        nil, 
    )
    if err != nil {
        return err
    }

    err = channel.Publish(
        "",
        queue.Name,
        false,
        false,
        amqp.Publishing{
            ContentType: 	"text/plain",
            Body:        	[]byte(messageBody),
			DeliveryMode:	amqp.Persistent,
        },
    )
    if err != nil {
        return err
    }

    return nil
}

File: src/daemon/gis-updater/utilis/rabbitmq.py
import pika
import time

class Rabbitmq:
    def __init__(self, vhost, user, password):
        self._vhost = vhost
        self._user = user
        self._password = password
        self._connection = None
        self.init()

    def init(self):
        try:
            conn = self.create_connection()
            self._connection = conn
        except Exception as e:
            raise Exception(e)
        
    
    def create_connection(self):
        conn_str = f"amqp://{self._user}:{self._password}@broker:5672/{self._vhost}"
        try:
            connection = pika.BlockingConnection(pika.URLParameters(conn_str))
            return connection
        except Exception as e:
            raise Exception(f"Error connecting to RabbitMQ: {e}")


    def consume_from_queue(self, queue_name, time_duration):
        start_time = time.time()
        messages = []

        try:
            channel = self._connection.channel()
            _ = channel.queue_declare(queue=queue_name, durable=False)

            def callback(ch, method, properties, body):
                message_body = body.decode('utf-8')
                messages.append(message_body)
                ch.basic_ack(delivery_tag=method.delivery_tag)

            channel.basic_consume(queue=queue_name, on_message_callback=callback)
        
            print("Waiting for messages...")
            try:
                while time.time() - start_time < time_duration:
                    channel.connection.process_data_events(time_duration - (time.time() - start_time))
            except KeyboardInterrupt:
                pass
            if len(messages) > 0:
                print("Done receiving messages.")

            channel.close()
            return messages
        except Exception as e:
            raise Exception(f"Error consuming from queue: {e}")